modin/pandas/dataframe.py

```diff
@@ -1388,60 +1409,152 @@ def merge(
+        suffixes : tuple of (str, str), default ('_x', '_y')
+            Suffix to apply to overlapping column names in the left and right side, respectively.
+            To raise an exception on overlapping columns use (False, False).
+        copy : bool, default True
+            If False, avoid copy if possible.
+        indicator : bool or str, default False
+            If True, adds a column to output DataFrame called "_merge" with information
+            on the source of each row. If string, column with information on source of each row
+            will be added to output DataFrame, and column will be named value of string.
+            Information column is Categorical-type and takes on a value of "left_only"
+            for observations whose merge key only appears in 'left' DataFrame,
+            "right_only" for observations whose merge key only appears in 'right' DataFrame,
+            and "both" if the observation’s merge key is found in both.
+        validate : str, optional
+            If specified, checks if merge is of specified type.
+            - 'one_to_one' or '1:1': check if merge keys are unique in both left and right datasets.
+            - 'one_to_many' or '1:m': check if merge keys are unique in left dataset.
+            - 'many_to_one' or 'm:1': check if merge keys are unique in right dataset.
+            - many_to_many' or 'm:m': allowed, but does not result in checks.
+        Returns
+        -------
+        DataFrame
+             A DataFrame of the two merged objects.
+        """
         if isinstance(right, Series):
             if right.name is None:
-                raise ValueError("Cannot merge a Series without a name")
+                raise ValueError("Can't merge a Series without a name")
             else:
                 right = right.to_frame()
         if not isinstance(right, DataFrame):
             raise ValueError(
-                "can not merge DataFrame with instance of type "
-                "{}".format(type(right))
+                "Can't merge DataFrame with instance of type " "{}".format(type(right))
             )
-        if left_index is False or right_index is False:
-            return self._create_or_update_from_compiler(
-                self._query_compiler.join(
+
+        if left_index and right_index:
+            return self.join(
+                right, how=how, lsuffix=suffixes[0], rsuffix=suffixes[1], sort=sort
+            )
+
+        if how in ["left", "inner"] and left_index is False and right_index is False:
+            result = self.__constructor__(
+                query_compiler=self._query_compiler.merge(
                     right._query_compiler,
                     how=how,
                     on=on,
                     left_on=left_on,
                     right_on=right_on,
                     left_index=left_index,
                     right_index=right_index,
-                    sort=sort,
+                    sort=False,
```

Comment:
In many implementations, it might make a difference which algorithms is used when `sort` is True vs False. Any of the sort handling should happen at the query compiler level, I believe.

Reply:
I agree, moved this into query compiler level.